Constitutional Committee Ratification

teodanciu opened the pull request:
# Description

Adds constitutional committee ratification as a mandatory check in `Ratify`. 
Closes https://github.com/input-output-hk/cardano-ledger/issues/3701
Closes https://github.com/input-output-hk/cardano-ledger/issues/3444

The structure of the code is very similar to drep` ratification, so we could abstract it. 

Missing: 
* ~Fix ConwayFeatures test~  EDIT: Fixed  
* voting no when the committee is too small  (https://github.com/input-output-hk/cardano-ledger/issues/3444) 
*  ~there is a contradiction between[ a comment on the issue](https://github.com/input-output-hk/cardano-ledger/issues/3701#issuecomment-1706790738) and (what I think I see in ) the spec~  EDIT: clarified
   > This means that expired and resigned committee members still contribute to the ratio calculation, despite that their votes do not count. 
* ~some unseemly duplication in the test, should find a better way to set expired or resigned members with less noise~ EDIT: refactored to remove the duplication!

Might be just a problem of semantics, but in this PR, I wouldn't say that expired and resigned "contribute to the ratio" (unless we interpret that as:  because they're expired or resigned we don't count them at all..)



<!-- Add your description here, if it fixes a particular issue please provide a
[link](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword=)
to the issue. -->

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated
- [ ] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [ ] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/input-output-hk/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [ ] Self-reviewed the diff

teodanciu committed:
Update `RatifyEnv` with `CommitteeState` field

in order to ratify committee votes

teodanciu committed:
Implement constitutional committee ratification

as a mandatory check in `Ratify` rule

teodanciu committed:
Add tests for committee ratification logic

teodanciu committed:
Rename `RatifySpec` to `DRepRatifySpec`

to make the specialization clear and for uniformity with
`CommitteeRatifySpec`

lehins commented:
> there is a contradiction betweenhttps://github.com/input-output-hk/cardano-ledger/issues/3701#issuecomment-1706790738 and (what I think I see in ) the spec:

Yeah, that comment on the ticket was wrong. I've fixed it. All expired and resigned CC members are considered as if they abstained:

https://github.com/input-output-hk/formal-ledger-specifications/blob/5feb58cd4f43aa331748503a2bfc9d2a04a1f015/src/Ledger/Ratify.lagda#L258-L261

teodanciu committed:
Fix bug causing the incorrect update of `CommitteeState` in Epoch

rule

teodanciu commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Epoch.hs:
```diff
@@ -372,6 +372,6 @@ instance
 updateCommitteeState :: StrictMaybe (Committee era) -> CommitteeState era -> CommitteeState era
 updateCommitteeState committee (CommitteeState creds) =
-  CommitteeState $ Map.difference creds members
+  CommitteeState $ Map.restrictKeys creds (Map.keysSet members)
```
`Map.difference` doesn't do the right thing here, but can't think of a better way to do this : only keep in `creds` the  keys that are in `members` .. Is there a better way?

lehins replied:
You are looking for `Map.intersection`

teodanciu replied:
Yes, great! didn't realize this worked for different value types of Map. Thanks!

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Ratify.hs:
```diff
@@ -103,10 +109,56 @@ instance
+    SNothing -> False -- this happens if we have no committee, in which case the committee vote is `no`
+    SJust r ->
+      -- short circuit on zero threshold.
+      -- This happen if the action is NoConfidence or UpdateCommittee, in which case the committee vote is 'yes'
```
The case you mention is actually impossible, so we wouldn't need to short circuit on it.
```suggestion
      -- This can happen when quorum is set to 0, then committee vote is always 'yes'
```

teodanciu committed:
Adjust `ConwayFeatures` test to take committee ratification into account

lehins commented on eras/conway/impl/test/Test/Cardano/Ledger/Conway/CommitteeRatifySpec.hs:
```diff
@@ -0,0 +1,330 @@
+      , let expire = (maybeToList . join) (Map.lookup ck csCommitteeCreds) `List.isInfixOf` hotCreds
+      ]
+   in
+    TestData (Map.fromList updatedMembers) votes committeeState
```
`elem` for lists is bad, even for tests. Also conversion to a list can be avoided
```suggestion
  let
    hotCredsSet = Set.fromList hotCreds
   in
    TestData members votes $ committeeState {
        csCommitteeCreds = 
          Map.map (\mhk -> mhk >>= \hk -> hk <$ guard (hk `Set.notMember` hotCredSet)) updatedState
      }
```

teodanciu replied:
Love this version, thank you

lehins commented on eras/conway/impl/test/Test/Cardano/Ledger/Conway/CommitteeRatifySpec.hs:
```diff
@@ -0,0 +1,330 @@
+    ( Map.union
+        (Map.fromList [(cred, VoteNo) | cred <- votedNo])
+        (Map.fromList [(cred, Abstain) | cred <- votedAbstain])
+    )
```
I suspect all of the tests will become slightly simpler if votes were actually sets, rather than lists
```suggestion
    Map.unions
      [ Map.fromSet (const VoteYes) votedYes
      , Map.fromSet (const VoteNo) votedNo
      , Map.fromSet (const Abstain) votedAbstain
      ]
```

teodanciu replied:
Great suggestion!  Many things are easier now, with the votes in Sets.

lehins commented on eras/conway/impl/test/Test/Cardano/Ledger/Conway/CommitteeRatifySpec.hs:
```diff
@@ -0,0 +1,330 @@
+  Gen (CommitteeState era)
+genNonResignedCommitteeState coldCreds = do
+  hotCreds <- vectorOf (length coldCreds) (arbitrary @(Credential 'HotCommitteeRole (EraCrypto era)))
+  pure $ CommitteeState $ Map.fromList $ zip coldCreds (Just <$> hotCreds)
```
Same here. It is better to operate directly on sets and maps and avoid lists.
Also, I think it would be a good idea to occasionally test a case when a different coldCred would map to the same hotCred (not an unreasonable situation when two CC member use the same script for some reason)
```suggestion
  Set (Credential 'ColdCommitteeRole (EraCrypto era)) ->
  Gen (CommitteeState era)
genNonResignedCommitteeState coldCreds = do
  hotCredsMap <- sequence $ Map.fromSet (\_ -> Just <$> arbitrary) coldCreds
  frequency
    [ (9, pure $ CommitteeState hotCredsMap)
    , (1, CommitteeState <$> overwriteWithDuplicate hotCredsMap)
    ]
  where
    overwriteWithDuplicate m
      | Map.size < 2 = m
      | otherwise = do
        fromIx <- choose (0, Map.size m - 1)
        toIx <- choose (0, Map.size m - 1)
        pure $ Map.updateAt (\_ _ -> Just $ Map.elemAt fromIx m) toIx m
```

teodanciu replied:
Hm this is interesting... The tests are failing if the committeeState contains duplicate values, so what you said: different cold keys mapping to the same hot key. 
The reason is that the votes are indexed by the hot key, so now instead of two votes, we'll have only 1..
I'm not sure yet if it's because of how the test data is generated - I'm sure that's at least one problem, but the the votes is still a `Map (Credential 'HotCommitteeRole (EraCrypto era)) Vote` .  
Is that what we expect, just one vote , if two cold keys map to the same hot key?

teodanciu replied:
For now I pushed a version that doesn't duplicate the hot keys, will change it back to your suggestion after figuring out the relationship of duplicated hot keys with the voting

lehins replied:
The only two ways it can work is one vote can be counted either once or at least once, when the same hot credential maps to many cold credentials. If I understand the spec correctly, and as I would implement it too, I would count it as many times as there are active cold credentials that match that hot credential.

So, I would suggest you try and make sure it works that way and test are passing with this too, but we should certainly discuss it on Monday meeting and see what @WhatisRT thinks about it, since I am sure he thought about this scenario.

teodanciu replied:
I adjusted the tests to support duplicate hotkeys. 
It presented some surprising challenges in terms of generating data, but it was fun to implement and I took the opportunity to also remove the duplication between generating expired and resigned members,. 
All in all I think it was a great suggestion and worth to have this case tested.  Thank you !

lehins commented on eras/conway/impl/test/Test/Cardano/Ledger/Conway/CommitteeRatifySpec.hs:
```diff
@@ -0,0 +1,330 @@
+  Era era =>
+  Gen [Credential 'ColdCommitteeRole (EraCrypto era)]
+genNonEmptyColdCreds =
+  listOf1 (arbitrary @(Credential 'ColdCommitteeRole (EraCrypto era)))
```
Type signature for generator will inferred, also producing a set right away will be better, since we never care about duplicate cold keys (and if I remember correctly duplicates won't actually be generated, not with any sort of reasonable probability that is)
```suggestion
genNonEmptyColdCreds ::
  Era era => Gen (Set (Credential 'ColdCommitteeRole (EraCrypto era))
genNonEmptyColdCreds = Set.fromList <$> listOf1 arbitrary
```

teodanciu committed:
Include check for minimum committee size in the threshold calculation

teodanciu committed:
Don't delay actions if the committee size is too small

instead rely on the threshold calculation to cause a "no" vote from the
committee in this case

lehins commented on eras/conway/impl/test/Test/Cardano/Ledger/Conway/CommitteeRatifySpec.hs:
```diff
@@ -0,0 +1,331 @@
+  committeeState@(CommitteeState {csCommitteeCreds}) <- genNonResignedCommitteeState @era coldCreds
+  let hotCreds = catMaybes $ Map.elems csCommitteeCreds
+      votes = distributeVotes @era ratios hotCreds
+      members = Map.fromList $ zip coldCreds (repeat (EpochNo 10))
```
It's usually better to generate values, rather than hard code specific ones in generators:
```suggestion
   members <- 
     Map.fromList . zip coldCreds <$> vectorOf (length coldCreds) (EpochNo <$> choose (10, 100))
```

lehins commented on eras/conway/impl/test/Test/Cardano/Ledger/Conway/CommitteeRatifySpec.hs:
```diff
@@ -0,0 +1,331 @@
+        size = fromIntegral $ length l
+        (xs, rest) = splitAt (ceiling (x * size)) l
+        (ys, rest') = splitAt (ceiling (y * size)) rest
+        (zs, rest'') = splitAt (ceiling (z * size)) rest'
```
I believe rounding will produce a bit more fare.
```suggestion
        (xs, rest) = splitAt (round (x * size)) l
        (ys, rest') = splitAt (round (y * size)) rest
        (zs, rest'') = splitAt (round (z * size)) rest'
```

lehins reviewed (approved):
Majority of my comments are about how the test suite can be improved. 

From what I can see nothing is blocking this PR from being merged.

Great work!!!

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/Governance.hs:
```diff
@@ -677,10 +683,16 @@ votingCommitteeThresholdInternal committee = \case
+            SJust t -> VotingThreshold t
+            SNothing -> NoVotingThreshold
+    minSize = fromIntegral $ pp ^. ppCommitteeMinSizeL
+    committeeSize = Map.size $ foldMap' committeeMembers committee
```
Now this is wrong.

`NoVotingThreshold` means that action will be accepted.

This is a wrong place to handle this case of committee too small

teodanciu replied:
Yes, I wasn't sure about this at all actually

lehins replied:
Ignore my comment, you are totally right!!!! :smile:

lehins reviewed (changes requested):
Spoke too soon :grin: 

~Latest changes made it wrong~ After looking a bit closer, they are actually fine.

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/Governance.hs:
```diff
@@ -677,10 +683,16 @@ votingCommitteeThresholdInternal committee = \case
+            SJust t -> VotingThreshold t
+            SNothing -> NoVotingThreshold
+    minSize = fromIntegral $ pp ^. ppCommitteeMinSizeL
+    committeeSize = Map.size $ foldMap' committeeMembers committee
```
```suggestion
    minSize = pp ^. ppCommitteeMinSizeL
    committeeSize = fromIntegral $ Map.size $ foldMap' committeeMembers committee
```

lehins commented on eras/conway/impl/CHANGELOG.md:
```diff
@@ -1,7 +1,10 @@
+* Add to `Ratify`:
+  * `committeeAccepted`
+  * `committeeAcceptedRatio`
+* Add `reCommitteeState` to `RatifyEnv``
```
```suggestion
* Add `reCommitteeState` to `RatifyEnv`
```

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/Governance.hs:
```diff
@@ -690,28 +691,32 @@ votingStakePoolThresholdInternal pp isElectedCommittee action =
-  isVotingAllowed . votingCommitteeThresholdInternal committee
+isCommitteeVotingAllowed :: ConwayEraPParams era => PParams era -> GovAction era -> Bool
+isCommitteeVotingAllowed pparams =
+  isVotingAllowed . votingCommitteeThresholdInternal pparams committee
```
Pparams don't actually affect if the voting by CC memebrs is allowed or not, so we can pass any value for this argument (same as with stake pools and dreps)
```suggestion
isCommitteeVotingAllowed :: GovAction era -> Bool
isCommitteeVotingAllowed =
  isVotingAllowed . votingCommitteeThresholdInternal def committee
```

teodanciu replied:
Nice simplification

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/Governance.hs:
```diff
@@ -720,10 +725,16 @@ votingCommitteeThresholdInternal committee = \case
+      | committeeSize < minSize = NoVotingThreshold
+      | otherwise =
+          case committeeQuorum <$> committee of
+            SJust t -> VotingThreshold t
```
Slight simplification that actually matches the comment a little better :wink: 
```suggestion
    threshold =
          case committeeQuorum <$> committee of
            -- if the committee size is smaller than the minimum given in pparams,
            -- we treat it as if we had no committee
            SJust t | committeeSize >= minSize -> VotingThreshold t
```

teodanciu replied:
Love it.  I didn't know you could "or" in a case match :grimacing:  Thank you

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Gov.hs:
```diff
@@ -221,16 +221,17 @@ noSuchGovActions govSnapshots gaIds =
 checkVotesAreValid ::
   forall era.
   ConwayEraPParams era =>
+  PParams era ->
```
No longer needed
```suggestion
```

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Ratify.hs:
```diff
@@ -255,12 +307,12 @@ ratifyTransition = do
         && withdrawalCanWithdraw gasAction
         && spoAccepted st env ast
         && dRepAccepted env st gas
+        && committeeAccepted st env gas
```
I think committee will be the fastest among SPOs and DReps, so I think we should do this first, thus short circuit earlier
```suggestion
        && committeeAccepted st env gas
        && spoAccepted st env ast
        && dRepAccepted env st gas
```

teodanciu replied:
Good point!

lehins reviewed (approved):
Looks great. Thank you for doing this work!
